testnet/stacks-node/src/burnchains/db_indexer.rs

```
@@ -0,0 +1,669 @@
+            }
+        };
+
+        up_cursor = cursor_header.parent_header_hash;
+    };
+
+    // Step 2: Set `is_canonical` to false from the old tip (inclusive) to the greatest
+    // common ancestor (exclusive).
+    let mut down_cursor = BurnchainHeaderHash(old_tip.header_hash());
+    loop {
+        let cursor_header = match query_row::<BurnHeaderDBRow, _>(
+            &transaction,
+            "SELECT * FROM headers WHERE header_hash = ?1",
+            &[&down_cursor],
+        )? {
+            Some(header) => header,
+            None => {
+                // TODO: Should this be an error?
+                panic!("Do we hit here?");
+            }
+        };
+
+        if cursor_header.header_hash == greatest_common_ancestor.header_hash {
+            break;
+        }
+
+        match transaction.execute(
+            "UPDATE headers SET is_canonical = 0 WHERE header_hash = ?1",
+            &[&down_cursor],
+        ) {
+            Ok(_) => {}
+            Err(e) => {
+                return Err(BurnchainError::DBError(db_error::SqliteError(e)));
+            }
+        };
+
+        down_cursor = cursor_header.parent_header_hash;
+    }
+
+    Ok(greatest_common_ancestor.height)
+}
+
+/// Returns the first ancestor of `last_canonical_tip` that is marked canonical. After a re-org, this
+/// can be used to find the greatest common ancestor between the new and old chain tips.
+fn find_first_canonical_ancestor(
+    connection: &DBConn,
+    last_canonical_tip: BurnchainHeaderHash,
+) -> Result<u64, BurnchainError> {
+    let mut cursor = last_canonical_tip;
+    loop {
+        let cursor_header = match query_row::<BurnHeaderDBRow, _>(
+            connection,
+            "SELECT * FROM headers WHERE header_hash = ?1",
+            &[&cursor],
+        )? {
+            Some(header) => header,
+            None => {
+                // TODO: Should this be an error?
+                panic!("Do we hit here?");
```

Yes, this should raise an error

---

changed to an Err